src/dds_data.py

```diff
@@ -10,208 +10,333 @@
  * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
 """
 
-from PySide6.QtCore import QObject, Signal, Slot
+from PySide6.QtCore import QObject, Signal, Slot, QThread, Qt
 from cyclonedds.builtin import DcpsEndpoint, DcpsParticipant
 import threading
 import logging
+import time
+import copy
+from queue import Queue
+from typing import Dict, List
+import gc
 
 from dds_service import builtin_observer
-from dds_qos import qos_match
+from dds_qos import qos_match, dds_qos_policy_id
 from utils import singleton, EntityType
 
 
+class DataEndpoint:
+    def __init__(self, endpoint: DcpsEndpoint, entity_type) -> None:
+        self.endpoint: DcpsEndpoint = endpoint
+        self.entity_type: EntityType = entity_type
+        self.participant = None
+        self.missmatches : Dict[str, List[dds_qos_policy_id]]= {}
+
+    def isReader(self):
+        return self.entity_type == EntityType.READER
+
+    def isWriter(self):
+        return self.entity_type == EntityType.WRITER
+
+    def link_participant(self, participant: DcpsParticipant):
+        if str(self.endpoint.participant_key) == str(participant.key):
+            self.participant = participant
+
+
+class DataTopic:
+    def __init__(self, name) -> None:
+        self.name = name
+        self.reader_endpoints: Dict[str, DataEndpoint] = {}
+        self.writer_endpoints: Dict[str, DataEndpoint] = {}
+
+    def add_endpoint(self, endpoint: DataEndpoint):
+        if endpoint.isReader():
+            if str(endpoint.endpoint.key) not in self.reader_endpoints:
+                self.reader_endpoints[str(endpoint.endpoint.key)] = endpoint
+        else:
+            if str(endpoint.endpoint.key) not in self.writer_endpoints:
+                self.writer_endpoints[str(endpoint.endpoint.key)] = endpoint
+
+        self.check_qos_mismatch(endpoint)
+
+    def remove_endpoint(self, endpointKey: str):
+        if endpointKey in self.reader_endpoints:
+            for mimKey in self.reader_endpoints[endpointKey].missmatches:
+                if mimKey in self.writer_endpoints:
+                    if endpointKey in self.writer_endpoints[mimKey].missmatches:
```

Comment:
yes you are right the `if` in line 65 is may a bit over guardedly. 
In the mismatches of a reader can only be writer ids.
Since we are working on a copy here there should be no need for a lock.

Good id to use the guid and not the generic string, i will do that in a separate PR: